### ISSUE TITLE
[export] Error loudly when tracing over a module not in the module hierarchy

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#118410]
Code like this:
```
inp = (torch.tensor([0.1, 0.1]),)
linear = torch.nn.Linear(2, 2)

class Module(torch.nn.Module):
    def forward(self, x):
      x = x + 1
      y = x.t()
      y = y.relu()
      y = linear(y)
      return y

 ep = export(Module(), inp)
```
has weird behavior today. In eager, `linear.weight` and `linear.bias` are not in `Module`'s state dict, but we detect them as parameters/buffers and lift them due to how we do input handling.

I think that export should just error loudly on oddly shaped code like this. In practical settings, it's almost never what you want, since things would not train correctly, and `load_state_dict` etc. would be totally broken.

**Comment by jansel** [comment]
IMO this will result in bad UX for people using `torch.compile()` and trying to migrate to export or AOTInductor.  This will work fine today, but the proposed change would break it.

I'd suggest just making it a warning.

**Comment by suo** [comment]
But like, will any user code ever do this? Training is totally broken if you do this right? Since state dict will be missing some weights.

**Comment by jansel** [comment]
You can do stuff like stitch two models together that were trained independently.  Or use it for fine-tuning only part of a model.

I think our general stance should be to default to supporting stuff. People use PyTorch in lots of weird ways.

**Comment by suo** [comment]
Oh sure yeah. To clarify—this will fail today with an obscure error message. To actually support it will require working through some design questions (how should we represent and serialize the state from an arbitrary number of unconnected modules?), so in the meantime we should at least have a clear message and way for the user to unblock themselves.

**Comment by suo** [comment]
yeah for export. Also for AOTInductor, not sure how it would work with weight update/runtime constant foldnig (which weights? how do we address them?), but maybe that works and I'm just not aware

**Comment by jansel** [comment]
Can we just error if you try to FX-serialize it then?  And if you don't try to do that it works fine.

**Comment by suo** [comment]
Hmmm, let me poke around and try. It's a little tricky now—export tries to assemble a data structure with strong invariants, and then the serialization code writes it out with minimal business logic.
